applications/api.py

```diff
@@ -0,0 +1,62 @@
+"""API for bootcamp applications app"""
+from django.db import transaction
+
+from applications.constants import AppStates, REVIEW_STATUS_REJECTED
+from applications.models import BootcampApplication
+from ecommerce.models import Order
+
+
+@transaction.atomic
+def get_or_create_bootcamp_application(user, bootcamp_run):
+    """
+    Fetches a bootcamp application for a user if it exists. Otherwise, an application is created with the correct
+    state.
+
+    Args:
+        user (User): The user applying for the bootcamp run
+        bootcamp_run (klasses.models.BootcampRun): The bootcamp run to which the user is applying
+
+    Returns:
+        BootcampApplication: The bootcamp application
+    """
+    bootcamp_app, created = (
+        BootcampApplication.objects
+        .select_for_update()
+        .get_or_create(
+            user=user,
+            bootcamp_run=bootcamp_run,
+        )
+    )
+    if created:
+        derived_state = derive_application_state(bootcamp_app)
```

Comment:
Is the user going to be able to create application for a bootcamp before he/she fills out the profile information?

Reply:
It _shouldn't_ be possible, but I added that logic into the function so that (a) it would return the correct derived state no matter what application object it received, and (b) it would catch any users that slipped through the cracks and somehow ended up with an incomplete profile. If we want to make the assumption that a user's profile must be filled out if they are logged in, we should get rid of the "awaiting profile" state entirely